packages/core/src/browser/widget-open-handler.ts

```diff
@@ -71,22 +72,50 @@ export abstract class WidgetOpenHandler<W extends BaseWidget> implements OpenHan
      */
     async open(uri: URI, options?: WidgetOpenerOptions): Promise<W> {
         const widget = await this.getOrCreateWidget(uri, options);
-        this.doOpen(widget, options);
+        await this.doOpen(widget, options);
         return widget;
     }
-    protected doOpen(widget: W, options?: WidgetOpenerOptions): void {
+    protected async doOpen(widget: W, options?: WidgetOpenerOptions): Promise<void> {
         const op: WidgetOpenerOptions = {
             mode: 'activate',
             ...options
         };
         if (!widget.isAttached) {
             this.shell.addWidget(widget, op.widgetOptions || { area: 'main' });
         }
+        const promises: Promise<void>[] = [];
         if (op.mode === 'activate') {
+            promises.push(this.onActive(widget));
+            promises.push(this.onReveal(widget));
             this.shell.activateWidget(widget.id);
         } else if (op.mode === 'reveal') {
+            promises.push(this.onReveal(widget));
             this.shell.revealWidget(widget.id);
         }
+        await Promise.all(promises);
+    }
+    protected onActive(widget: W): Promise<void> {
+        return new Promise(resolve => {
+            const listener = (shell: ApplicationShell, args: FocusTracker.IChangedArgs<Widget>) => {
+                if (args.newValue === widget) {
+                    this.shell.activeChanged.disconnect(listener);
+                    resolve();
+                }
+            };
+            this.shell.activeChanged.connect(listener);
```

Comment:
could it happen, that widget is already active and `activeChange` is not emitted?

Reply:
I know, on blur of widget, shell won't say this widget is active.

Reply:
Maybe it's bike-shedding, and I only can think of a synthetic edge case where some flawed command (maybe triggered via key binding) would try to open same widget as the currently active one.

Reply:
I will add the guard to check whether the widget is already active. It also can be a problem if a  widget does not focus anything but then it does not play nicely with the shell as well.

Reply:
thanks!

Reply:
I've added guards